### ISSUE TITLE
[Issue]: large bf16 inputs leads to nan

### ISSUE THREAD

**Issue by xinyazhang** [ROCm/aotriton#54]
### Problem Description

This is due to the core logic of FA algorithm:
https://github.com/ROCm/aotriton/blob/f6b28a9b7265b69e3df54ea6ba0237e8a8d6f736/tritonsrc/fwd_kernel_inner.py#L95-L97

`qk_scale * tl.max(qk, 1))` uses `vmul`, but `qk * qk_scale - m_ij[:, None]` uses `vfma`
`vfma` has higher precision than `vmul`+`vadd` because it only round once.

More specifically, Suppose the vectors all only contain one element, and `qk = round(133120.0 * 133120.0) = 17720934400.0`, `qk_scale = round(0.25 * 1.44269502162933349609) = 0.360673755407333374023`
Then
```
qk * qk_scale - m_ij
= qk * qk_scale - qk_scale * qk
= fma(qk, qk_scale, -vmul(qk, qk_scale))
= fma(qk, qk_scale, -round(17720934400.0, 0.360673755407333374023))
= round(qk * qk_scale - 6391475712.0)
= round(17720934400.0 * 0.360673755407333374023 - round(17720934400.0, 0.360673755407333374023))
= round(6391475959.3749999999922470912 - 6391475712.0)
= 247.375
```
Therefore, `p` in the code above yields to `exp2(247.375)=inf`. This consequently leads to nan in following steps.

### Operating System

N/A

### CPU

N/A

### GPU

MI300X

### ROCm Version

ROCm 6.2.3

**Comment by AIwork4me** [comment]
**Primary verification stack (all data below):** AMD official `torch 2.12.0+rocm7.14.0` wheel (HIP 7.14.60850), bundled `libaotriton_v2.so.0.11.2`, gfx1151 (Radeon 8060S, Strix Halo), `TORCH_ROCM_AOTRITON_ENABLE_EXPERIMENTAL=1`. Secondary: TheRock `2.10.0+rocm7.13.0a20260513` — same failure envelope (both stacks predate the fix branch).

## TL;DR

Issue #54's NaN/wrong-output failure is reachable from the **stock SDPA path on shipped wheels**, with a precise and narrow trigger envelope, and the existing fix branch `fix/gfx1151-bf16-dot-fp32-accumulator` (commit `8232d69672`, 2026-06-01, "fix(gfx1151): explicit out_dtype=tl.float32 in fwd tl.dot") is **still unmerged** (main and every tag incl. `0.11.2b`…`0.13.50tp` carry the unfixed `qk += (Qk_scale * tl.dot(q0, k0))`). We hit this in production serving a 27B VLM judge and lost a day to it; please merge + release.

## Exact trigger envelope (measured, 7.14.0 wheel)

bf16, batch 1, 16 heads, non-causal, MATH backend as reference:

| variable | result |
|---|---|
| head_dim **72**, seq **>1024** (1025 suffices) | **wrong output / NaN** |
| head_dim 64, 80, 88, 96, 104, 112, 120, 128 (any seq) | ok |
| head_dim 72, seq ≤ 1024 | ok |
| head_dim 72, seq 2048, **is_causal=True** | **ok** (same shape!) |
| head_dim 72, seq 2048, fp16 | wrong (dtype-independent) |
| batch 1..16 × heads 2..32 (incl. GQA 16q/2kv) | wrong — independent of grid occupancy |
| direct `aten::_scaled_dot_product_flash_attention` | reproduces → inside the flash kernel, not SDPA dispatch |

Per-query-position error is scattered across the whole sequence and the **logsumexp itself is wrong** — the softmax denominator is corrupted, not just PV (single-run snapshot: 8/16 blocks of 128 bad at seq=2048, LSE rel up to 2.1e-1; a re-run showed 10/16 bad blocks — magnitudes drift, see below). Same tensors, 5 repeated calls, one process: rel err [3.9e-1, nan, 5.5e-1, 1.4e-1, 3.4e-1] — magnitudes drift run-to-run and NaN appears ~1/5. hd=72 selects the BLOCK_DMODEL-split dot path (`BLOCK_DMODEL1 > 0`) and non-causal long-seq picks the tile config the fix commit calls out ("BLOCK_M = 128 and BLOCK_N = 64"); causal at the same shape selects a different config and is spared — consistent with the fix comment.

## Impact

Any model whose attention has head_dim 72 and >1024 tokens, non-causal, on gfx1151 with AOTriton enabled. Concrete production case: the **Qwen-Image-Bench Q-Judger (Qwen3.5 27B VLM)** ViT is 16 heads × 1152 hidden = head_dim 72; a 1024×1024 image is 4096 patches → judge generations degenerate to repeated-`!` garbage under vLLM with AOTriton on, and are correct with it off (observed live during serving; the raw broken outputs were overwritten by later re-runs and are not archived — the envelope above is the reproducible form). Text-decoder paths (head_dim 64/128) are unaffected, which is why LLM-only users don't see it.

## Repro (copy-paste)

```python
import torch, torch.nn.functional as F  # run with TORCH_ROCM_AOTRITON_ENABLE_EXPERIMENTAL=1

def check(seq, hd=72, causal=False):
    torch.manual_seed(0)
    q, k, v = (torch.randn(1, 16, seq, hd, device="cuda", dtype=torch.bfloat16) for _ in range(3))
    with torch.nn.attention.sdpa_kernel(torch.nn.attention.SDPBackend.MATH):
        ref = F.scaled_dot_product_attention(q, k, v, is_causal=causal)
    out = F.scaled_dot_product_attention(q, k, v, is_causal=causal)
    nan = torch.isnan(out).any().item()
    rel = float((out.float() - ref.float()).abs().max() / ref.float().abs().max())
    print(f"hd={hd} seq={seq} causal={causal}: nan={nan} rel={rel:.2e}",
          "BAD" if (nan or rel > 0.05) else "ok")

check(1024)          # ok
check(1025)          # BAD
check(2048)          # BAD
check(2048, causal=True)  # ok — same shape, different tile config
check(2048, hd=64)   # ok
```

## Ask

1. Merge `fix/gfx1151-bf16-dot-fp32-accumulator` (8232d69672) into main and include it in the next release; the unmerged branch also carries `test_issue54_gfx1151_bf16_nan.py`.
2. Consider widening that test's coverage: it exercises NaN on large bf16 inputs; our envelope (hd=72/seq≥1025/**non-causal only**, plus scattered wrong-but-finite outputs and LSE corruption) is a strict superset worth pinning.
3. Optionally: the fix branch's own TODO notes the INT8 path has the same implicit-out_dtype issue (`tl.dot(q,k).to(tl.float32)` casts after the fact) — worth fixing in the same pass.

Workaround for affected users: keep `TORCH_ROCM_AOTRITON_ENABLE_EXPERIMENTAL` unset, or force the MATH backend for head_dim-72 modules.
